File: haskell/kindergarten-garden/Garden.hs
module Garden
    (
      Plant(..),
      garden,
      defaultGarden,
      lookupPlants
    ) where

    import Data.List
    import Data.List.Split
    import Data.Maybe
    import qualified Data.Map.Strict as M

    type Garden = M.Map Children [Plant]
    type Children = String

    data Plant = Violets | Grass | Clover | Radishes deriving (Show, Eq)

    garden :: [Children] -> String -> Garden
    garden children g = M.fromList $ zip children' plants
      where
            children' = sort children
            plants = fmap (fmap toPlant . concat) by4
            by4 = transpose by2
            by2 = fmap (chunksOf 2) plantsRows
            plantsRows = splitOn "\n" g
            toPlant c = case c of
              'V' -> Violets
              'G' -> Grass
              'C' -> Clover
              'R' -> Radishes


    defaultGarden :: String -> Garden
    defaultGarden = garden defaultChildren

    defaultChildren :: [String]
    defaultChildren = ["Alice", "Bob", "Charlie", "David", "Eve", "Fred",
      "Ginny", "Harriet", "Ileana", "Joseph", "Kincaid", "Larry"]

    lookupPlants :: String -> Garden -> [Plant]
    lookupPlants child = fromJust . M.lookup child

File: elixir/kindergarten-garden/garden.exs
defmodule Garden do
  @students [
          :alice,
          :bob,
          :charlie,
          :david,
          :eve,
          :fred,
          :ginny,
          :harriet,
          :ileana,
          :joseph,
          :kincaid,
          :larry
        ]

  @plants %{
    ?R => :radishes,
    ?C => :clover,
    ?V => :violets,
    ?G => :grass
  }

  @spec info(String.t(), list) :: map
  def info(info_string, student_names) do
    students = Enum.sort(student_names)
    seed = String.to_charlist(String.replace(info_string, "\n", ""))
    Map.new(students, fn student -> {
      student,
      fetch_plants(seed, offset(students, student))
    } end)
  end
  def info(info_string), do: info(info_string, @students)

  defp fetch_plants(seed, offset) do
    case plants(seed, offset) do
        {a,b,c,d} -> {a,b,c,d}
        _ -> {}
    end
  end

  defp plants(seed, offset) do
    first2(seed, offset) ++ last2(seed, offset)
    |> Enum.map(&Map.get(@plants, &1))
    |> List.to_tuple()
  end

  defp first2(seed, offset), do: Enum.slice(seed, offset * 2, 2)
  defp last2(seed, offset) do
    Enum.slice(seed, div(Enum.count(seed), 2) + (2 * offset), 2)
  end

  defp offset(students, name) do
    Enum.find_index(students, fn x -> x == name end)
  end
end
